Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -4860,68 +4860,89 @@ - (void)reloadViewItems
         previousViewItemTimestamp = viewItem.interaction.timestampForSorting;
     }
 
-    // Update the "shouldShowDate" property of the view items.
+    // Update the properties of the view items.
     //
-    // First iterate in reverse order.
-    OWSInteractionType lastInteractionType = OWSInteractionType_Unknown;
-    MessageReceiptStatus lastReceiptStatus = MessageReceiptStatusUploading;
-    NSString *_Nullable lastIncomingSenderId = nil;
-    for (ConversationViewItem *viewItem in viewItems.reverseObjectEnumerator) {
+    // NOTE: This logic uses shouldShowDate which is set in the previous pass.
```

Comment:
Reworked this into a single pass as you suggested and tweaked behavior a bit to fix some edge cases.